### PR TITLE
Feature/dataclass schema defaults

### LINKED ISSUE #968: dataclass field doesn't change required schema
# Bug

Please complete:
* OS: Mac OS
* Python version `import sys; print(sys.version)`: **3.6.7** with dataclasses backport
* Pydantic version `import pydantic; print(pydantic.VERSION)`: **1.0**

It is known that `@pydantic.dataclasses.dataclass` doesn't fully mimic the `BaseModel`. However, we'd like use `.schema()` representation faithfully. I'm wondering what's the best workaround.

Adding a `Config` for `b` with `required=False` doesn't carry over to `.schema()` either.

```py
from typing import Dict

from dataclasses import field
from pydantic import Field, ValidationError
from pydantic.dataclasses import dataclass

@dataclass
class D1:
    a: int
    # b: Dict[str, int] = {"x":10} # mutable default, not allowed
@dataclass
class D2:
    a: int
    b: Dict[str, int] = Field({"x":10}) # pydantic.Field, unsafe and invalid default
@dataclass        
class D3:
    a: int
    b: Dict[str, int] = field(default_factory=lambda: {"x":10}) # std lib, safe and invalid default
        
class C:
    fields = {"b": { "required": False}} # doesn't carry over!!

@dataclass(config=C)        
class D4:
    a: int
    b: Dict[str, int] = field(default_factory=lambda: {"x":10})

print(D1(10))
try:
    print(D2(10))
except ValidationError as e:
    print(e) # pydantic/dataclasses.py:77

print(D3(10))
req = D3.__pydantic_model__.schema()["required"]
if req != ["a"]:
    print('only ["a"] should be required, is:', req)
    
print()
print(D4.__pydantic_model__.schema()["required"]) # same as D3
```

Output
```
D1(a=10)
1 validation error for D2
b
  value is not a valid dict (type=type_error.dict)
D3(a=10, b={'x': 10})
only ["a"] should be required, is: ['a', 'b']

['a', 'b']
```

### PR DESCRIPTION
**Pull request by ahirner**
## Change Summary
* Make use of default_factory lambda when creating pydantic fields
* Add separate test for this kind of default
* Add field for this kind of default in test_schema

## Related issue number
Fixes #968 
If fine and complete, lets discuss docs

## Checklist

* [X] Unit tests for the changes exist
* [X] Tests pass on CI and coverage remains
* [X] Documentation reflects the changes where applicable
* [X] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: docs/examples/dataclasses_default_schema.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+from pydantic.dataclasses import dataclass
+
+@dataclass
+class User:
+    id: int
+    name: str = 'John Doe'
+    friends: List[int] = dataclasses.field(default_factory=lambda: [0])
+
+user = User(id='42')
+print(user.__pydantic_model__.schema()['properties']['friends'])
```
Comment (samuelcolvin) [inline docs/examples/dataclasses_default_schema.py]:
```suggestion
print(user.__pydantic_model__.schema())
```
Is probably enough, best to keep documentation as simple as possible while demonstrating the point.

Comment (ahirner) [reply docs/examples/dataclasses_default_schema.py]: I had this at first then thought the prints use up too much space. So you prefer the complete schema for sure?

Comment (samuelcolvin) [reply docs/examples/dataclasses_default_schema.py]: yes please.

**Comment by samuelcolvin** [comment]
great, thank you very much.

**Pull request merged**